### ISSUE TITLE
nullable checking doesn't go far enough to check non-nullable variables being set to null

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#32359]
The following code snippet shows a situation where a non-nullable string gets inadvertently assigned null.
Limitations like this mean that we can't trust the nullable warnings - which makes the feature less useful as people will then not trust it and will abandon trying to fix their code.

To reproduce, paste the following code into a C# console project, and set the project to use the C#8 compiler with #nullable enable set.

Note that there's no warning produced, but the string gets set to null.

			string str;
			{
				using ( RegistryKey rk = Registry.CurrentUser.OpenSubKey( &quot;Software&quot; ) )
				{
					// No warning, but the cast from a null object will result in null
					str = (string) rk.GetValue( &quot;WontExist&quot; );
				}
			}
			if ( str == null )
			{
				Console.WriteLine( &quot;It is null&quot; );
			}

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/412355/nullable-checking-doesnt-go-far-enough-to-check-no.html
VSTS ticketId: 754373_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by jmarolf** [comment]
Note it will warn on `as`
```C#
string str;
using (var rk = Registry.CurrentUser.OpenSubKey("Software"))
{
    // Warning	CS8600	Converting null literal or possible null value to non-nullable type.
    str = rk.GetValue("WontExist") as string;
}
if (str == null)
{
    Console.WriteLine("It is null");
}
```
I am torn.
I feel like a direct cast is basically telling the type system "I know this will do the right thing". 
On the other hand there are a lot of APIs in .NET that return objects and you are expected to cast them to the underlying type.  

Seems like if I get a type from a null oblivious library I should assume that the returned object _can_ be null and require a null check before assignment.

**Comment by Dave-Lowndes** [comment]
> I am torn.
> I feel like a direct cast is basically telling the type system "I know this will do the right thing".

I had the same thought too with my example, but it was a concise example of a real issue I hit when testing. Using 'as' didn't seem quite the same situation as it can inherently result in null.

> On the other hand there are a lot of APIs in .NET that return objects and you are expected to cast them to the underlying type.

Yes.

>Seems like if I get a type from a null oblivious library I should assume that the returned object can be null and require a null check before assignment.

Yes, I feel that it should assume the object is nullable - which is the opposite of what's currently assumed. I presume the C# team have a good reason for doing that?

**Comment by CyrusNajmabadi** [comment]
> I presume the C# team have a good reason for doing that?

Yes.  because you casted directly to a type.  So it seems odd to say: well, it's not that type.  Furthermore, it would make it much more verbose to then have to have some other mechnaism that says "no, damnit, it's the non-nullable version of this type".

As it stands, tehre are two supported forms:

1. `(Type)`.  Casts to the type, expects the user knows what they're doing.
2. `as Type`.  Casts to the type, but may be `null`.

This gives hte user two good ways of handling this situation.  if `(Type)` was the same as `as Type` we'd both have redundancy and we'd need an entirely different way to express this very desirable operation.  That is far worse than just having two separate forms to support to the two common desired ops.

**Comment by Dave-Lowndes** [comment]
> Yes.  because you casted directly to a type.

Ignore the fact the example shows a cast. I was trying to keep things simple. Assume instead that it's an external method that is defined to return a string, and in some circumstance it returns a null.

For example, if you have a class library project built with C#7 with this code:
```C#
namespace ClassLibrary
{
    public class Class1
    {
		public static string IReturnNull()
		{
			return null;
		}
    }
}
```
Then in the C#8 nullable enabled project you call it like this:
```C#
string myString = ClassLibrary.Class1.IReturnNull();
```
Your C#8 project then has a non-nullable string which is null, and there's no warning :(

**Comment by jmarolf** [comment]
> Assume instead that it's an external method that is defined to return a string, and in some circumstance it returns a null.

This is a general critique of how null obliviousness is handled today.  The fundamental assumption of the compiler it that: 
- `T` implies the intention is the type _cannot_ be null
- `T?` implies the intention is the type _can_ be null

For libraries that are null oblivious when given a `T` the compiler can assume:
1. That the type is not null
2. That the type is allowed to be null
3. That the type is some third state (not either 1 or 2 until proven otherwise, forcing the user to disambiguate)

Of these options, 3 cannot work in my opinion.  Users are not in the position to really reason about the intent of an arbitrary method they are calling in all but the most trivial use case.  In addition if we went with 3 what would the concrete type of `var` be below?  `T` or `T?`
```C#
var x = NullObliviousMethodReturningT();
```

Option 2 feels like it might be correct as we are saying "I am unable to prove this is non-null so you need to do something" but that goes against the fundamental assumption that nullable reference types is based on.  Nullable reference types says "Since runtime behavior assumes that references are not null (as it is a runtime exception for them to be otherwise) the type system can also assume that they are non-null". The hope is that for the 90% case nulls are not allowed on reference types and this assumption is correct for the users code.  If instead we had all null-oblivious types treated as `T?` this would quickly get confusing.  Users would see type declarations of `T` and say "I can't know what the semantics are for this type without also knowing the compiler options and pragmas for this file". You then effectively have two languages, each with completely different semantics.

This leaves us with option 1.  But I still think we need to do something to make this better.  I think an analyzer that I turn on that warns me when a null-oblivious type is being used as a non-null type would be useful.  I could then suppress this warning using the normal analyzer suppression mechanisms in my code for things that I feel are safe oblivious to non-null conversions or add nullability annotations for types that I am unsure about. I could also just not opt-into this analyzer at all if I didn't want to be bothered.

**Comment by ewlloyd** [comment]
Just spitballing from the cheap seats here...Would it be possible for the `#nullable enable` pragma to decorate the output assembly with a `NullableRefAttribute`, then have the compiler use it to handle null-oblivious assemblies differently (assume `T == T?`)?

**Comment by jmarolf** [comment]
@ewlloyd 	The design today is that the compiler will emit an attribute for assemblies that are compiled with the nullable option.  This is how null-obliviousness is expected to be determined. We are still left with the question of what the compiler is supposed to do with this information. 

Using the sematics you have suggested if `GetValue()` is defined in a null-oblivious library then the below code will result in a warning.
```C#
#nullable enable
T a = GetValue();
```
However if the user goes to the definition of the method they may see something like this
```C#
T GetValue => new T();
```
Where `T` is clearly not `T?` even though it is being treated that way.  I worry that this would make it very difficult for newcomers to reason about which types are expected to be nullable and which are not.

**Comment by ewlloyd** [comment]
> @ewlloyd The design today is that the compiler will emit an attribute for assemblies that are compiled with the nullable option. This is how null-obliviousness is expected to be determined. We are still left with the question of what the compiler is supposed to do with this information.
> 
I had a feeling it was already there, but I just wanted to level-set from my own naive perspective -- I haven't spent a lot of time exploring the intricacies of Nullable Reference Types (more's the pity).

> Using the sematics you have suggested if `GetValue()` is defined in a null-oblivious library then the below code will result in a warning.
> 
> ```cs
> #nullable enable
> T a = GetValue();
> ```
> 
> However if the user goes to the definition of the method they may see something like this
> 
> ```cs
> T GetValue => new T();
> ```
> 
> Where `T` is clearly not `T?` even though it is being treated that way. I worry that this would make it very difficult for newcomers to reason about which types are expected to be nullable and which are not.

I definitely see your point here, but hear me out. Let's consider adding another pragma that implements foreign `T == T?` semantics (let's call it `#nullable strict` for now). It's okay to do without this if you're writing end-user-facing code, and can still write defensively against the odd `NullReferenceException`.

However, if you use `#nullable enable` without `#nullable strict` in a class-library assembly, then you really can't make any guarantees of null-safety to your callers, either, and really have no business advertising yourself as null-safe. Otherwise, the whole Nullable Reference Types feature takes on a decidedly screen-door-on-a-submarine vibe.

In fact, when you really think about it, the `#nullable strict` isn't such a hot idea after all, because users who are unaware of it will get a false sense of security (as noted by the OP). Perhaps the better solution is to just have the compiler force the strict semantics by default, and issue more explicit messages. Even if you don't have the room to spell it out completely in the message, the user's just a Google (okay, Bing) search away from true enlightenment.

**Comment by jmarolf** [comment]
We have now officially steered into [csharplang ](https://github.com/dotnet/csharplang) territory.  @gafter 	can you move this issue over there please?
